### PR TITLE
test: extend test coverage

### PR DESCRIPTION
**Pull request by TanguyDeTaxis**
### Summary

This PR aims to increase test coverage across consensus, cryptographic, and types components, focusing on security-critical paths and BFT invariants. The tests include both traditional unit/integration tests and property-based tests using `proptest`.

**Note:** Not everything here may be worth merging. Feel free to cherry-pick commits that seem relevant and discard the rest.

**+2,870 lines** | **27 files changed** | **7 commits**

---

### Build & Test Optimization (Latest Commit)

In [PR #516](https://github.com/Telcoin-Association/telcoin-network/pull/516), I mentioned that my benchmarks showed negligible gains from build optimizations, leading me to remove the `config.toml`. However, I later discovered that my local configuration had cached old settings and the changes were not actually being applied.

After properly testing, the results are dramatically different: **nextest reduces test execution time from several minutes to just a few seconds** (on my machine). The commit adds incremental compilation, sccache support, and nextest configuration for parallel test execution.

---

### Coverage Report

I generated a coverage report for the workspace. See attached full report in comment.

| Metric | Coverage |
|--------|----------|
| **Function** | 72.77% (2157/2964) |
| **Line** | 73.11% (17835/24394) |
| **Region** | 70.44% (27544/39103) |

Generated with (excluding some flaky tests):

```bash
cargo llvm-cov --html --workspace --exclude tn-network-libp2p -- \
  --skip test_layereddb_contains_key \
  --skip test_layereddb_iter_reverse \
  --skip test_layereddb_multi_insert \
  --skip test_layereddb_multi_remove \
  --skip test_layereddb_skip
```

---

#### 🔐 Types (`crates/types/tests/it/`)

**Certificate Validation Tests** (`certificate_tests.rs`)
- Quorum requirement validation (2f+1 signatures)
- BLS aggregate signature verification
- Epoch mismatch detection
- Genesis certificate validation
- Unknown authority rejection
- Empty key set verification (quorum edge case)
- Quorum threshold formula verification for N=1 to 100

**Committee Property Tests** (`committee_props.rs`)
- Membership invariants (authority lookup, voting power)
- Quorum threshold properties (majority requirement, BFT bounds)
- Two quorums intersection guarantee (BFT liveness)
- Deterministic leader election

**Serialization Tests** (`serde_tests.rs`)
- Round-trip serialization for Header, Certificate, Vote
- Bincode compatibility verification

---

#### ⚡ Consensus (`crates/consensus/primary/`)

**Bullshark Commit Tests** (`tests/it/bullshark_commit_tests.rs`)
- f+1 leader support requirement
- Recursive commit behavior
- sub_dag_index sequential ordering
- Leader election determinism

**DAG State Tests** (`tests/it/dag_state_tests.rs`)
- Equivocation detection (same author, same round, different cert)
- Parent verification (missing parents rejected)
- Garbage collection boundary enforcement
- Idempotent certificate insertion

**Leader Schedule Tests** (`tests/it/leader_schedule_tests.rs`)
- Round-robin leader coverage
- Bad node swapping mechanics
- Deterministic swap table generation

**Recovery Tests** (`tests/it/recovery_tests.rs`)
- sub_dag_index persistence across restarts
- Epoch boundary recovery
- State consistency after crash

**Consensus Properties** (`tests/it/consensus_props.rs`)
- Leader determinism (same round = same leader)
- No consecutive leaders property
- Swap table determinism

**DAG Integrity Tests** (`src/consensus/tests/consensus_tests.rs`)
- Missing parent rejection
- Missing parent round rejection
- GC boundary enforcement (`round <= gc_round` rejected)
- Equivocation detection at ConsensusState level
- Duplicate insertion idempotence

**Network Equivocation Tests** (`src/tests/network_tests.rs`)
- Vote caching (same digest returns cached response)
- Equivocation rejection (different digest, same round)
- Older round rejection (already voted for later round)
- Per-authority equivocation tracking

---

#### 🔄 Executor (`crates/consensus/executor/tests/it/`)

- Output number monotonicity (strictly increasing)
- Batch fetching validation (all batches included)

---

#### 🔗 State Sync (`crates/state-sync/`)

**Epoch Validation** (`src/epoch.rs` - unit tests)
- `epoch_committee_valid` function tests
- Equal committee validation
- Smaller/larger committee handling
- 2/3 quorum requirement for epoch records
- Boundary condition tests

**Digest Chain Tests** (`tests/it/main.rs`)
- ConsensusOutput persistence
- Parent hash chain validation
- Lookup by hash (digest index)
- Digest determinism
- Collision resistance (different inputs → different digests)
- Fork detection (wrong parent_hash detected)

---

#### 💰 Economics (`crates/tn-reth/tests/it/`)

**Fee Calculation Properties** (`economics_props.rs`)
- Gas penalty monotonicity
- Zero penalty at target
- Maximum penalty bounds
- Quadratic scaling verification

---

### Test Infrastructure

- Added `proptest` dependency for property-based testing
- Added `.cargo/config.toml` for test configuration
- Test utilities for deterministic BLS key generation

---

### Security Properties Verified

| Property | Test Location |
|----------|--------------|
| 2f+1 quorum requirement | `certificate_tests.rs` |
| BLS signature verification | `certificate_tests.rs` |
| Equivocation detection | `dag_state_tests.rs`, `consensus_tests.rs`, `network_tests.rs` |
| DAG parent integrity | `consensus_tests.rs`, `dag_state_tests.rs` |
| GC boundary enforcement | `consensus_tests.rs` |
| Fork detection | `state-sync/tests/it/main.rs` |
| Leader determinism | `consensus_props.rs` |
| Two quorums intersect | `committee_props.rs` |

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -147,9 +175,9 @@ revert-submodule:
 
 # workspace tests that don't require faucet credentials
 public-tests:
-	cargo test --workspace --exclude tn-faucet --no-fail-fast -- --show-output ;
-	cargo test -p telcoin-network --test it test_epoch_boundary -- --ignored ;
-	cargo test test_restarts -- --ignored ;
+	cargo nextest run --workspace --exclude tn-faucet --no-fail-fast ;
+	cargo nextest run -p telcoin-network --test it --filter 'test_epoch_boundary' --run-ignored all ;
+	cargo nextest run --filter 'test_restarts' --run-ignored all ;
```
Comment (grantkee) [inline Makefile]: Small tweak to nextest commands:
```bash
error: unexpected argument '--filter' found

# fixed
cargo nextest run -p telcoin-network --test it --run-ignored all test_epoch_boundary ;

cargo nextest run --run-ignored all test_restarts ;
```

File: Cargo.toml [hunk 2]
```diff
@@ -30,6 +30,7 @@ members = [
     "crates/execution/faucet",
     "crates/execution/tn-rpc",
 ]
+exclude = ["fuzz"]
```
Comment (grantkee) [inline Cargo.toml]: Leftover?

File: .cargo/config.toml
Comment (grantkee) [inline .cargo/config.toml]: Thanks for working on build optimizations! The profile settings for debug, split-debuginfo, opt-level, and codegen-units look good and should help with compile times.

  I have a concern about the `rustc-wrapper` and `rustflags` settings:
```toml
  [build]
  rustc-wrapper = "sccache"

  [target.x86_64-unknown-linux-gnu]
  rustflags = ["-C", "link-arg=-fuse-ld=mold"]
```

These require contributors to have sccache and mold installed, otherwise the build fails immediately:
```bash
  error: could not execute process `sccache ... rustc` (never executed)
  Caused by: No such file or directory (os error 2)
```

This adds friction for new contributors and anyone cloning the repo for the first time. The CI also has to remove this config file to avoid these issues (in `pr.yaml`).

I'd prefer to remove the `[build]` and `[target.x86_64-unknown-linux-gnu]` section from the project config, and instead document these as optional local setup recommendations in `CONTRIBUTING.md`:

---
  #### Optional: Faster builds

  For faster compile times, add these to your personal `~/.cargo/config.toml`:

  **sccache** (all platforms):
  ```toml
  [build]
  rustc-wrapper = "sccache"
  ```
  Install: `cargo install sccache --locked`

  **mold linker** (Linux only):
  ```toml
  [target.x86_64-unknown-linux-gnu]
  rustflags = ["-C", "link-arg=-fuse-ld=mold"]
  ```
  Install: `sudo apt install mold` or see https://github.com/rui314/mold

---

This way developers who want the speedup can opt-in, while the project builds out-of-the-box for everyone else.

That being said, I'm using `sccache` and `mold` now and think it's a great suggestion!

Comment (TanguyDeTaxis) [reply .cargo/config.toml]: Didn't see it like that at first time but totally agreed! Removed the [build] and [target] sections. The config now only has dev/test profile optimizations (debuginfo, opt-level, codegen-units) which don't require any external tools.

File: crates/tn-reth/tests/it/economics_props.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    /// Lower usage percentage should result in higher or equal penalty (monotonicity).
+    /// As gas efficiency decreases, the penalty should increase.
+    #[test]
+    fn prop_penalty_increases_with_lower_usage(
+        gas_limit in MIN_GAS_LIMIT_THRESHOLD * 2..30_000_000u64,
+        usage_pct_high in 1u64..USAGE_THRESHOLD_PCT,
+        usage_delta in 1u64..10
+    ) {
+        let usage_pct_low = usage_pct_high.saturating_sub(usage_delta).max(1);
+        prop_assume!(usage_pct_low < usage_pct_high);
```
[the remark this answers is not in the document]
Comment (TanguyDeTaxis) [reply crates/tn-reth/tests/it/economics_props.rs]: Nice catch! It fixes the edge case you identified. I went one step further though — even with high starting at 2 and a separate low range, prop_assume!(low < high) would still 
  discard cases where the two values happen to be equal. So I used a delta-based approach that makes every generated case valid by construction:                                                                                    
                                                                                                                                                                                                                                    
  - usage_pct_high in 2..USAGE_THRESHOLD_PCT (your suggestion)                                                                                                                                                                      
  - usage_delta in 1..10                                              
  - usage_pct_low = usage_pct_high.saturating_sub(usage_delta).max(1)

  Since high >= 2 and delta >= 1, low is always strictly less than high — no prop_assume! needed, zero wasted test cases. That said, I'm not 100% sure it is what you expected — let me know if you spot a scenario where this breaks down or if it's a totally wrong way.
